dao/authentication_dao.go

```diff
@@ -315,15 +315,54 @@ func (a *authenticationDaoImpl) Delete(uid string) (*m.Authentication, error) {
 	for _, key := range keys {
 		if strings.HasSuffix(key, uid) {
 			path := fmt.Sprintf("secret/metadata/%d/%s", *a.TenantID, key)
-			sec, err := Vault.Delete(path)
+			// fetch the key (for returning)
+			sec, _ := a.getKey(key)
+			// ...then delete it
+			_, err := Vault.Delete(path)
 
-			return authFromVault(sec), err
+			return sec, err
 		}
 	}
 
 	return nil, util.NewErrNotFound("authentication")
 }
 
+// Deletes all authentications for a parent resource. Sources are handled
+// differently than other parent resources mostly due to the cascade effect from
+// a source delete
+func (a *authenticationDaoImpl) Cleanup(resourceType string, resourceID int64) error {
+	resourceType = util.Capitalize(resourceType)
+
+	auths, _, err := a.List(100, 0, []util.Filter{})
+	if err != nil {
+		return err
+	}
+
+	for i := range auths {
+		if resourceType == "Source" {
+			if auths[i].SourceID != resourceID {
+				continue
+			}
+
+			_, err := a.Delete(auths[i].ID)
+			if err != nil {
+				return err
+			}
+		} else {
+			if auths[i].ResourceID != resourceID || auths[i].ResourceType != resourceType {
```

Comment:
maybe 

```go
... 
}else if auths[i].ResourceID == resourceID && auths[i].ResourceType == resourceType {
...
```

Reply:
also it would possible to merge both branches, like:
```go
resourceCondition := auths[i].ResourceID == resourceID && auths[i].ResourceType == resourceType
sourceCondition  := resourceType == "Source" && auths[i].SourceID == resourceID

if resourceCondition || sourceCondition {
...
```
names are just suggestion but maybe It could improve readability.

Reply:
I reworked them to be more readable. I also added a DB implementation since we do not have a vault backend yet.